library/core/src/ptr/volatile.rs

```diff
@@ -0,0 +1,240 @@
+/// to not be elided or reordered by the compiler across other volatile
+/// operations.
+///
+/// # Notes
+///
+/// Rust does not currently have a rigorously and formally defined memory model,
+/// so the precise semantics of what "volatile" means here is subject to change
+/// over time. That being said, the semantics will almost always end up pretty
+/// similar to [C11's definition of volatile][c11].
+///
+/// The compiler shouldn't change the relative order or number of volatile
+/// memory operations. However, volatile memory operations on zero-sized types
+/// (e.g., if a zero-sized type is passed to `read_volatile`) are noops
+/// and may be ignored.
+///
+/// [c11]: http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `src` must be [valid] for reads.
+///
+/// * `src` must be properly aligned.
+///
+/// * `src` must point to a properly initialized value of type `T`.
+///
+/// Like [read], `read_volatile` creates a bitwise copy of `T`, regardless of
+/// whether `T` is [`Copy`]. If `T` is not [`Copy`], using both the returned
+/// value and the value at `*src` can [violate memory safety][read-ownership].
+/// However, storing non-[`Copy`] types in volatile memory is almost certainly
+/// incorrect.
+///
+/// Note that even if `T` has size `0`, the pointer must be properly aligned.
+///
+/// [valid]: crate::ptr#safety
+/// [read-ownership]: crate::ptr::read#ownership-of-the-returned-value
+/// [read]: crate::ptr::read
+///
+/// Just like in C, whether an operation is volatile has no bearing whatsoever
+/// on questions involving concurrent access from multiple threads. Volatile
+/// accesses behave exactly like non-atomic accesses in that regard. In particular,
+/// a race between a `read_volatile` and any write operation to the same location
+/// is undefined behavior.
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// let x = 12;
+/// let y = &x as *const i32;
+///
+/// unsafe {
+///     assert_eq!(std::ptr::read_volatile(y), 12);
+/// }
+/// ```
+#[inline]
+#[stable(feature = "volatile", since = "1.9.0")]
+#[cfg_attr(miri, track_caller)] // Even without panics, this helps for Miri backtraces
+#[rustc_diagnostic_item = "ptr_read_volatile"]
+pub unsafe fn read_volatile<T>(src: *const T) -> T {
+    // SAFETY: the caller must uphold the safety contract for `volatile_load`.
+    unsafe {
+        crate::ub_checks::assert_unsafe_precondition!(
+            check_language_ub,
+            "ptr::read_volatile requires that the pointer argument is aligned and non-null",
+            (
+                addr: *const () = src as *const (),
+                align: usize = align_of::<T>(),
+                is_zst: bool = T::IS_ZST,
+            ) => crate::ub_checks::maybe_is_aligned_and_not_null(addr, align, is_zst)
+        );
+        cfg_match! {
```

Comment:
Of course ideally all of this magic would happen in the intrinsic / in LLVM. *If* we're going to implement volatile semantics in lib code for any appreciable amount of time, though, it should probably be done in sys PAL modules rather than inline due to the quantity of different platforms which will likely want to provide `asm!` volatile semantics.

Reply:
`sys` is a std thing, we can't use it here. And anyway it's for abstracting OS/platform differences, not CPU architecture differences.